### ISSUE TITLE
[bug] useProvider isn't returning the correct provider when using Hardhat locally

### ISSUE THREAD

**Issue by thecooltechguy** [tmm/wagmi#411]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.3.2

### Current Behavior

When you connect your Metamask wallet to a local hardhat node (running on localhost:8545), and then connect this wallet to a React app using wagmi, the output of the `useProvider()` hook returns a provider connected to `homestead`, instead of the local hardhat node.

Furthermore, `useContractRead` & `useToken` also use the default `homestead`-connected provider (instead of a provider connected to the local hardhat node), whereas `useContractWrite` correctly uses a provider/signer connected to the local hardhat node.

### Expected Behavior

When you connect a Metamask wallet (connected to a localhost hardhat node) with wagmi:

1) The `useProvider` hook should return a provider connected to this local hardhat node, and not a provider connected to `homestead`.

2) Hooks like `useToken` & `useContractRead` should use a provider connected to the local hardhat node, and not one connected to `homestead`.

### Steps To Reproduce

I've created a mini repro example showing this behavior at: https://codesandbox.io/s/cool-platform-0u6n2z?file=/src/index.tsx

Basically, once you connect a Metamask wallet (that's connected to a local hardhat node), you'll see that the `useProvider()` hook returns a provider that's connected to `homestead` instead.

If Codesandbox shows an error with importing `import { InjectedConnector } from "wagmi/connectors/injected";`, simply comment out this import line (line 3 @ App.tsx), save file (cmd/ctrl+s), uncomment this import line, save file (cmd/ctrl+s), and everything should work again.

### Link to Minimal Reproducible Example (CodeSandbox, StackBlitz, etc.)

https://codesandbox.io/s/cool-platform-0u6n2z?file=/src/index.tsx

**Comment by thecooltechguy** [comment]
Just wanted to add that I've updated the code repro example to also use the suggested workaround from https://github.com/tmm/wagmi/issues/71 (ie., to define the client's provider function that returns an explicit localhost provider for Hardhat's chain id), but this still doesn't resolve the bug.

**Comment by tmm** [comment]
Working for me. What does the Hardhat chain look like in MM?

<img width="334" alt="image" src="https://user-images.githubusercontent.com/6759464/167047301-814bd0cc-af70-4786-bea7-a7f864083ae0.png">

**Comment by L-TheG** [comment]
@thecooltechguy I am running into a very similar issue #425 
Did you manage to find a solution?

**Comment by thecooltechguy** [comment]
@L-TheG I'm also facing this issue but unfortunately couldn't find a solution yet. Also, do the contractWrite functions work for you? Or, is it only the contractRead + other "read"-based functions (like useToken, etc.) that don't work for you? For me, the contract write functions work, but not the contract read ones (since they use mainnet instead of hardhat).

@tmm I verified that my Hardhat chain configuration looks exactly like you posted within MM.

**Comment by thecooltechguy** [comment]
After some debugging, it appears that the root cause for this bug is that inside the `useChainId` hook (https://github.com/tmm/wagmi/blob/8a5f6c209049c35f16f9c6a86cd4ea71106e84a6/packages/react/src/hooks/utils/useChainId.ts#L7), it returns `provider.network.chainId` -- however, `provider.network` is `undefined`. As a quick test, I edited the `useChainId` hook to simply return the input `chainId` without querying the `provider` via `useProvider`, and my code began to work again (ie., read-based smart contract calls worked, etc.).

From looking at the `ethers` docs & code, it appears that the network for a provider can be obtained via `provider.getNetwork()` (instead of `provider.network`), which returns a promise: https://github.com/ethers-io/ethers.js/blob/01aea705ce60b1c42d2f465b162cb339a0e94392/packages/providers/src.ts/base-provider.ts#L1135

@tmm So perhaps, should the `useChainId` hook be updated to fetch the network via the provider's `getNetwork()` function? Or, am I maybe misunderstanding something here?

**Comment by RafaelVidaurre** [comment]
Same here, weird that this hasn't been addressed?

**Comment by tmm** [comment]
@thecooltechguy thanks for the additional info. One thing I noticed that is missing from the minimal reproducible example is setting up a connector with Hardhat. By default, `createClient` defaults to use `InjectedConnector` with the `defaultChains` configured. You should probably configure it to use Hardhat as well:

```ts
const client = createClient({
  connectors: [new InjectedConnector({ chains: [chain.hardhat] })
})
```

@RafaelVidaurre Other parts of wagmi are taking priority and since I work on wagmi in my free time, this hasn't been addressed yet.

**Comment by RafaelVidaurre** [comment]
> @thecooltechguy thanks for the additional info. One thing I noticed that is missing from the minimal reproducible example is setting up a connector with Hardhat. By default, `createClient` defaults to use `InjectedConnector` with the `defaultChains` configured. You should probably configure it to use Hardhat as well:
> 
> ```ts
> const client = createClient({
>   connectors: [new InjectedConnector({ chains: [chain.hardhat] })
> })
> ```
> 
> @RafaelVidaurre Other parts of wagmi are taking priority and since I work on wagmi in my free time, this hasn't been addressed yet.

Thanks for the reply, and for maintaining this awesome library. Already got the hardhat chain added to the client on my end. I'll try patching this based on @thecooltechguy 's input and maybe submit a PR

**Comment by thecooltechguy** [comment]
@tmm @RafaelVidaurre Thanks for sharing your comments!

@tmm What do you think of the following approach for using `getNetwork` inside `useChainId` w/ `react-query` (since the function is async):

```
import { GetProviderNetworkArgs } from '@wagmi/core'
import { providers } from 'ethers'
import { QueryFunctionArgs } from '../../types'
import { useProvider } from '../providers'
import { useQuery } from './useQuery'

export type UseChainIdArgs = {
  chainId?: number
}

export const queryKey = ({
  provider,
}: Partial<GetProviderNetworkArgs>) => [{ entity: 'providerNetwork', provider }] as const

const queryFn = ({
  queryKey: [{ provider }],
}: QueryFunctionArgs<typeof queryKey>) => {
  if (!provider || !provider.getNetwork) {
    throw new Error('provider is invalid')
  }
  return provider.getNetwork();
}

export function useChainId({ chainId }: UseChainIdArgs = {}) {
  const provider = useProvider({ chainId })

  if (!provider || !provider.getNetwork) {
    throw new Error('provider is invalid')
  }

  const {data} = useQuery(queryKey({ provider }), queryFn);
  return (data as providers.Network)?.chainId;
}
```

**Comment by jamiew** [comment]
@thecooltechguy do you have a fork with that patch in it yet? I ran into this issue today and would be happy to give it a spin and help debug

p.s. props for the detailed issue reports you've filed here and thank you to @tmm for writing and sharing wagmi!

**Comment by thecooltechguy** [comment]
@tmm @jamiew I've created a fork w/ this fix here: https://github.com/thecooltechguy/wagmi/tree/fix_issue_411

I tested this fork in my project and my "Contract Read" functions are now working again. 

However, I tried running the project's existing tests (by running `ALCHEMY_ID=... pnpm test:ci`) and it seems like some tests are now failing (due to various reasons, like timeouts, etc.). 

Also, the test for `useChainId` is also failing, which I still have to debug.

**Comment by jamiew** [comment]
@thecooltechguy fwiw switching to the (not yet released) `configureChains` API instead of `createClient` solved all my issues: https://github.com/tmm/wagmi/pull/408

Rainbowkit has some live examples of it as well

**Comment by jxom** [comment]
@thecooltechguy – can you confirm you are still running into these issues with `^0.4.0`?

**Comment by thecooltechguy** [comment]
@jamiew Thanks for sharing this! Using the new API you mentioned + Rainbowkit has solved the issue for me.

@jxom I'm no longer running into these issues so I'll close this issue. Thanks!
